File: 100/131 - Palindrome Partitioning.java
public class Solution {
    public List<List<String>> partition(String s) {
        boolean[][] isPalindrome;
        List<List<String>> ret = new ArrayList<List<String>>();
        int i, j;
        
        if (s == null)
            return ret;
        isPalindrome = new boolean[s.length()][s.length()];
        
        for (i = s.length() - 1; i >= 0; --i) {
            for (j = i; j < s.length(); ++j) {
                if (s.charAt(i) == s.charAt(j) && (i >= j - 2 || isPalindrome[i + 1][j - 1]))
                    isPalindrome[i][j] = true;
            }
        }
        
        helper(s, 0, isPalindrome, new ArrayList<String>(), ret);
        
        return ret;
    }
    
    private void helper(String s, int from, boolean[][] isPalindrome, List<String> prefix, List<List<String>> ret) {
        int i;
        String sub;
        
        if (from == s.length()) {
            ret.add(new ArrayList<String>(prefix));
            return;
        }
        
        for (i = from; i < s.length(); ++i) {
            if (isPalindrome[from][i]) {
                sub = s.substring(from, i + 1);
                prefix.add(sub);
                helper(s, i + 1, isPalindrome, prefix, ret);
                prefix.remove(prefix.size() - 1);
            }
        }
    }
}

File: csharp/131.palindrome-partitioning.cs
/*
 * @lc app=leetcode id=131 lang=csharp
 *
 * [131] Palindrome Partitioning
 */

using System.Collections.Generic;

public class Solution {

    private void PartitionDFS(string s, IList<IList<string>> ret, bool[,] isPalindrome, int i, int n, IList<string> curr) {
        if (i == n) {
            ret.Add(new List<string>(curr));
        }
        for (int j = i; j < n; ++j) {
            if (!isPalindrome[i, j]) {
                continue;
            }
            curr.Add(s.Substring(i, j - i + 1));
            PartitionDFS(s, ret, isPalindrome, j + 1, n, curr);
            curr.RemoveAt(curr.Count - 1);
        }
    }

    public IList<IList<string>> Partition(string s) {
        var ret = new List<IList<string>>();
        int n = s.Length;
        if (n == 0) {
            return ret;
        }
        bool[,] isPalindrome = new bool[n, n];
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j <= i; ++j) {
                if (s[i] == s[j] && (j >= i - 2 || isPalindrome[j + 1, i - 1])) {
                    isPalindrome[j, i] = true;
                }
            }
        }
        PartitionDFS(s, ret, isPalindrome, 0, n, new List<string>());
        return ret;
    }
}

